### PR TITLE
Donate CIDR-aligned ranges instead of just slicing at the mid-point.

### PR DESCRIPTION
**Pull request by bboreham**
If, say, we wanted to use the ranges for routing rules, having them aligned so you can describe each one with a single CIDR is useful.
Also fix up space tests to match.

### CODE REVIEW

File: net/address/address_test.go [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+	require.Equal(t, NewRange(2, 2), NewRange(1, 4).BiggestPow2AlignedRange())
+	require.Equal(t, NewRange(4, 4), NewRange(1, 8).BiggestPow2AlignedRange())
+	require.Equal(t, NewRange(4, 4), NewRange(2, 8).BiggestPow2AlignedRange())
+	require.Equal(t, NewRange(4, 4), NewRange(4, 8).BiggestPow2AlignedRange())
+	require.Equal(t, NewRange(8, 4), NewRange(8, 7).BiggestPow2AlignedRange())
+	require.Equal(t, NewRange(8, 8), NewRange(8, 8).BiggestPow2AlignedRange())
+	require.Equal(t, NewRange(0, 0x4000), NewRange(0, 0x7fff).BiggestPow2AlignedRange())
+	require.Equal(t, NewRange(0x2000, 2), NewRange(0x2000, 3).BiggestPow2AlignedRange())
+	require.Equal(t, NewRange(0xffff, 1), NewRange(0xffff, 1).BiggestPow2AlignedRange())
+}
```
Comment (rade) [inline net/address/address_test.go]: I would be tempted to write a helper that allows me to write

    check(t, 0x2000, 2, 0x2000, 3)

Also, I do rather struggle with working out why *exactly* these tests are done.

Comment (bboreham) [inline net/address/address_test.go]: > _exactly_ these tests

Something like "the first time each bit in the result changes or might change if there was a bug, plus some boundary conditions"

Comment (brb) [inline net/address/address_test.go]: Have you considered using https://golang.org/pkg/testing/quick/ ?

File: ipam/space/space.go [hunk 2]
```diff
@@ -122,8 +122,11 @@ func (s *Space) biggestFreeRange(r address.Range) (biggest address.Range) {
 	biggestSize := address.Count(0)
 	s.walkFree(r, func(chunk address.Range) bool {
 		if size := chunk.Size(); size >= biggestSize {
```
Comment (brb) [inline ipam/space/space.go]: The biggest chunk does not guarantee that it contains the biggest CIDR-aligned range.

File: ipam/space/space.go [hunk 3]
```diff
@@ -122,8 +122,11 @@ func (s *Space) biggestFreeRange(r address.Range) (biggest address.Range) {
...
-			biggest = chunk
-			biggestSize = size
+			chunk = chunk.BiggestPow2AlignedRange()
```
Comment (brb) [inline ipam/space/space.go]: The biggest chunk does not guarantee that it contains the biggest CIDR-aligned range.

If we don't care too much about a few CPU cycles being wasted, then I'd suggest to merge the if-statements.

Comment (bboreham) [inline ipam/space/space.go]: The first if-statement discards chunks that cannot possibly contain the biggest CIDR-aligned range.
The second if-statement accepts the biggest CIDR-aligned range.

Yes, it is coded this way to run faster than if we looked for the biggest CIDR-aligned chunk in each free chunk.

File: net/address/address.go [hunk 4]
```diff
@@ -35,6 +35,35 @@ func (r Range) AsCIDRString() string {
 	return CIDR{Addr: r.Start, PrefixLen: prefixLen}.String()
 }
 
+// return the highest bit set in a
```
Comment (brb) [inline net/address/address.go]: "Return the highest bit set in v"?

File: net/address/address.go [hunk 5]
```diff
@@ -35,6 +35,35 @@ func (r Range) AsCIDRString() string {
...
+	v |= v >> 1
+	v |= v >> 2
+	v |= v >> 4
+	v |= v >> 8
+	v |= v >> 16
+	// Cast to 64-bit to avoid overflow when original v >= 0x80000000
+	return Count((uint64(v) + 1) / 2)
+}
+
+func (r Range) BiggestPow2AlignedRange() Range {
```
Comment (brb) [inline net/address/address.go]: `BiggestCIDRRange`?

File: net/address/address_test.go [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+	for ; x > 1; x /= 2 {
+		if x&1 != 0 {
+			return false
+		}
+	}
+	return true
+}
+
+func TestBiggestPow2AlignedRange(t *testing.T) {
+	require.Equal(t, NewRange(0, 1), NewRange(0, 1).BiggestPow2AlignedRange())
```
Comment (brb) [inline net/address/address_test.go]: The `cover` tool says that the following five assertions do not execute all four possible execution paths. I'd suggest to have a test case per each.

File: net/address/address.go (hunk 5 above)
Comment (bboreham) [inline net/address/address.go]: I steered away from "CIDR" because we have a type called `CIDR` that we aren't using here, but `BiggestCIDRRange` might be ok.

Comment (bboreham) [inline net/address/address.go]: Should I paste in the Greek-letter explanation of what is going on here?

**Comment by bboreham** [comment]
I addressed most comments; PTAL

File: net/address/address.go (hunk 5 above)
Comment (brb) [inline net/address/address.go]: Ghm. I'm fine with understanding what it does from the code. What @rade @awh thinks?
